flat_map: Optimize conversion of SPL tuple to/from dictionary (named tuple)

Involved operator code: `com.ibm.streamsx.topology.functional.python/FlatMap`

Primitive operator does provide different code block for various types (python object/dict/named tuple) like source, map, aggregate.

current code results in following error when using type hints (named tuple as output style)

```
src/operator/flatten_flat_map.cpp: In member function \u2018virtual void SPL::_Operator::flatten_flat_map$OP::process(const SPL::Tuple&, uint32_t)\u2019:
src/operator/flatten_flat_map.cpp:130:41: error: \u2018SPL::_Operator::flatten_flat_map_Base::OPort0Type\u2019 has no member named \u2018get___spl_po\u2019
           pySplValueFromPyObject(otuple.get___spl_po(), item);
                                         ^
make: *** [build/operator/flatten_flat_map.o] Error 1
```

--- a/.gitignore
+++ b/.gitignore
@@ -1390,6 +1390,7 @@ test/python/spl/testtkpy_pip_toolkit/opt/.splpy/**
 test/python/topology/symlinkmodule.py
 test/python/topology/test_symlink_to_module
 test/python/topology/test.out
+test/python/topology/flat_map.txt
 test/python/topology/test1.test_TopologyLowLatency_JobConfig.json
 test/python/topology/test2_pending.test_pure_loop_JobConfig.json
 test/spl/testtk/impl/java/bin/testjava/MyPipe$StreamsModel.class

--- a/build.xml
+++ b/build.xml
@@ -114,7 +114,7 @@
     <delete dir="test/python/spl/testtkpy/com.ibm.streamsx.topology.pytest.pyec/PyTestMetrics"/>
     <delete dir="test/python/spl/testtkpy/com.ibm.streamsx.topology.pytest.pyec/PyTestOperatorContext"/>
     <delete>
-      <fileset dir="test/python/topology" includes="*.sab,*.json"/>
+      <fileset dir="test/python/topology" includes="*.sab,*.json,*.out,*.txt"/>
     </delete>
     <delete>
       <fileset dir="test/python/spl/tests" includes="*.sab,*.json"/>

--- a/com.ibm.streamsx.topology/CHANGELOG.md
+++ b/com.ibm.streamsx.topology/CHANGELOG.md
@@ -1,6 +1,10 @@
 # Changes
 ==========
 
+## master/latest
+
+* [#2486](https://github.com/IBMStreams/streamsx.topology/issues/2486) Support type hints with Python `flat_map`
+
 ## v1.15.1
 * [#2484](https://github.com/IBMStreams/streamsx.topology/issues/2484) Document known issues in Python doc
 * [#2478](https://github.com/IBMStreams/streamsx.topology/issues/2478) Support additional fields like `pipPackages` and `rpms` be provisioned for the "edge" image creation

--- a/com.ibm.streamsx.topology/com.ibm.streamsx.topology.functional.python/FlatMap/FlatMap_cpp.cgt
+++ b/com.ibm.streamsx.topology/com.ibm.streamsx.topology.functional.python/FlatMap/FlatMap_cpp.cgt
@@ -2,8 +2,15 @@
 
 @include "../pyspltuple.cgt"
 <%
+ my $pyoutstyle = splpy_tuplestyle($model->getOutputPortAt(0));
+%>
+
+<%
  # Select the Python wrapper function
  my $pywrapfunc= $pystyle_fn . '_in__pickle_iter';
+ if ($pyoutstyle eq 'dict') {
+   $pywrapfunc= $pystyle_fn . '_in__object_iter';
+ }
 %>
 
 // Default case is pass by pickled value in which case
@@ -17,6 +24,9 @@
 MY_OPERATOR::MY_OPERATOR() :
    funcop_(NULL),
    pyInStyleObj_(NULL),
+ <%if ($pyoutstyle eq 'dict') {%>    
+   pyOutNames_0(NULL),
+ <%}%>
    occ_(-1)
 { 
     const char * wrapfn = "<%=$pywrapfunc%>";
@@ -58,6 +68,10 @@ MY_OPERATOR::MY_OPERATOR() :
 #if SPLPY_OP_STATE_HANDLER == 1
    this->getContext().registerStateHandler(*this);
 #endif
+
+<%if ($pyoutstyle eq 'dict') {%>
+    pyOutNames_0 = Splpy::pyAttributeNames(getOutputPortAt(0));
+<%}%> 
 }
 
 MY_OPERATOR::~MY_OPERATOR() 
@@ -108,11 +122,26 @@ void MY_OPERATOR::process(Tuple const & tuple, uint32_t port)
       // construct spl blob and tuple from pickled return value
       OPort0Type otuple;
 
+    <%if ($pyoutstyle eq 'dict') {%>     
+      if (item == NULL)
+          return;
+      if (PyTuple_Check(item)) {
+          fromPyTupleToSPLTuple(item, otuple);
+          Py_DECREF(item);
+      } else if (PyDict_Check(item)) {
+          fromPyDictToSPLTuple(item, otuple);
+          Py_DECREF(item);
+      } else {
+    	  pySplValueFromPyObject(otuple.get_<%=$model->getOutputPortAt(0)->getAttributeAt(0)->getName()%>(), item);
+    	  Py_DECREF(item);
+      }      
+    <% } else { %>
       SPLPY_OUT_TUPLE_FLAT_MAP_BY_REF(otuple.get___spl_po(), item, occ_)
       {
           pySplValueFromPyObject(otuple.get___spl_po(), item);
           Py_DECREF(item); 
       }
+    <%}%>
       output_tuples.push_back(otuple);
     }
     Py_DECREF(pyIterator);
@@ -132,5 +161,24 @@ void MY_OPERATOR::process(Punctuation const & punct, uint32_t port)
 {
    forwardWindowPunctuation(punct);
 }
+
+<%
+  if ($pyoutstyle eq 'dict') {
+	my %cpp_tuple_types;  
+    # In this case we don't want the function that
+    # converts the Python tuple to an SPL tuple to
+    # copy attributes from the input port
+    my $iport;
+
+    my $oport = $model->getOutputPortAt(0);
+    my $oport_submission = 0;
+    my $otupleType = $oport->getSPLTupleType();
+    my @onames = SPL::CodeGen::Type::getAttributeNames($otupleType);
+    my @otypes = SPL::CodeGen::Type::getAttributeTypes($otupleType);
+
+%>
+// Create member function that converts Python tuple to SPL tuple
+@include  "../../opt/python/codegen/py_pyTupleTosplTuple.cgt"
+<%}%>
 
 <%SPL::CodeGen::implementationEpilogue($model);%>

--- a/com.ibm.streamsx.topology/com.ibm.streamsx.topology.functional.python/FlatMap/FlatMap_h.cgt
+++ b/com.ibm.streamsx.topology/com.ibm.streamsx.topology.functional.python/FlatMap/FlatMap_h.cgt
@@ -11,6 +11,11 @@ using namespace streamsx::topology;
 
 @include "../../opt/python/codegen/py_disallow_cr_trigger.cgt"
 
+@include "../pyspltuple.cgt"
+<%
+ my $pyoutstyle = splpy_tuplestyle($model->getOutputPortAt(0));
+%>
+
 class MY_OPERATOR : public MY_BASE_OPERATOR
 #if SPLPY_OP_STATE_HANDLER == 1
  , public SPL::StateHandler
@@ -30,6 +35,13 @@ public:
 #endif
 
 private:
+  <%
+  if ($pyoutstyle eq 'dict') {
+  %>
+    void fromPyTupleToSPLTuple(PyObject *pyDict, OPort0Type & otuple);
+    void fromPyDictToSPLTuple(PyObject *pyTuple, OPort0Type & otuple);
+  <%}%>
+  
     SplpyOp * op() { return funcop_; }
 
     // Members
@@ -37,7 +49,9 @@ private:
     SplpyFuncOp *funcop_;
     
     PyObject *pyInStyleObj_;
-
+  <%if ($pyoutstyle eq 'dict') {%>    
+	PyObject *pyOutNames_0;
+  <%}%>
     // Number of output connections when passing by ref
     // -1 when cannot pass by ref
     int32_t occ_;

--- a/com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/topology.py
+++ b/com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/topology.py
@@ -1606,6 +1606,20 @@ class Stream(_placement._Placement, object):
         and mapped stream corresponding to the input tuple
         that caused the exception.
 
+        Example: For a list of dict the ``flat_map`` emits **n** tuples for each input tuple received, with **n** the number of elements in the list::
+            
+            from typing import Iterable, List, NamedTuple
+
+            class SampleSchema(NamedTuple):
+                id: str
+                flag: bool
+
+            def flatten_dict(tpl) -> Iterable[SampleSchema]:
+                return tpl
+
+            # list_stream is a stream of list from dict as Python object, for example [{'id': '0', 'flag':True}]       
+            sample_stream = list_stream.flat_map(flatten_dict) # sample_stream is a named tuple stream of SampleSchema
+
         Returns:
             Stream: A Stream containing flattened and mapped tuples.
         Raises:

--- a/python/build.xml
+++ b/python/build.xml
@@ -34,7 +34,7 @@
     <delete>
      <fileset dir="${sphinx.dir}/source"
        includes="*.rst"
-       excludes="index.rst service.rst pysetup.rst"/>
+       excludes="index.rst service.rst pysetup.rst restrictions.rst"/>
    </delete>
   </target>
 </project>

--- a/test/python/build.xml
+++ b/test/python/build.xml
@@ -65,6 +65,7 @@
 		     <arg value="test2_parallel_key_partitioned.TestDistributedParallel"/>
 		     <arg value="test2.TestDistributedTopologyMethodsNew"/>
 		     <arg value="test2_aggregate.TestDistributedAggregate"/>
+		     <arg value="test2_flat_map.TestDistributedFlatMap"/>
 		     <arg value="test2_checkpoint.TestDistributedCheckpointing"/>
 		     <arg value="test2_consistent.TestDistributedConsistentRegion"/>
 		     <arg value="test2_ec.TestDistributedEc"/>

--- a/test/python/topology/test2_flat_map.py
+++ b/test/python/topology/test2_flat_map.py
@@ -1,0 +1,173 @@
+# Licensed Materials - Property of IBM
+# Copyright IBM Corp. 2020
+import unittest
+import itertools
+import os
+from streamsx.topology.topology import *
+from streamsx.topology.schema import CommonSchema, StreamSchema
+from streamsx.topology.tester import Tester
+from streamsx.topology.context import submit, ContextTypes, ConfigParams, JobConfig
+import streamsx.spl.op as op
+import typing
+from typing import Iterable, List, NamedTuple
+
+"""
+Test flat_map function with various input and output types.
+"""
+
+class NumbersSchema(NamedTuple):
+    num: int
+
+class NumbersListSchema(NamedTuple):
+    numlist: List[int]
+
+class GenerateNumbersNamedTuple(object):
+
+    def __init__(self, iterations, range):
+        self.iterations = iterations
+        self.range = range
+
+    def __call__(self) -> Iterable[NumbersListSchema]:
+        for num in itertools.count(1):
+            if num >  self.iterations:
+                break
+            l = []
+            l.extend(range(0, self.range))
+            yield {"numlist": l}
+
+class GenerateNumbers(object):
+
+    def __init__(self, iterations, range):
+        self.iterations = iterations
+        self.range = range
+
+    def __call__(self):
+        for num in itertools.count(1):
+            if num >  self.iterations:
+                break
+            l = []
+            l.extend(range(0, self.range))
+            yield {"numlist": l}
+
+class GenerateListOfDict(object):
+
+    def __init__(self, iterations, range):
+        self.iterations = iterations
+        self.range = range
+
+    def __call__(self):
+        for num in itertools.count(1):
+            if num >  self.iterations:
+                break
+            l = []
+            count = 0
+            while count < self.range:
+                d = {'num': count, "a":'a'+str(count), 'b':True}
+                l.append(d.copy())
+                count += 1
+            yield l
+
+def flatten_flat_map(tpl):
+    return tpl.numlist
+
+def flatten_flat_map_with_type_hints(tpl) -> Iterable[NumbersSchema]:
+    return tpl.numlist
+
+def flatten_flat_map_obj(tpl):
+    return tpl["numlist"]
+
+def flatten_flat_map_obj_with_type_hints(tpl) -> Iterable[NumbersSchema]:
+    return tpl["numlist"]
+
+class SampleSchema(NamedTuple):
+    num: int
+    a: str
+    b: bool
+
+def flatten_flat_map_dict_with_type_hints(tpl) -> Iterable[SampleSchema]:
+    return tpl
+
+#############################
+debug_flat_map_output = False
+#############################
+
+class TestFlatMap(unittest.TestCase):
+
+    def setUp(self):
+        Tester.setup_standalone(self)
+
+    def _runTest(self, topo, s):
+        if debug_flat_map_output:
+            import streamsx.standard.files as files
+            from streamsx.standard import CloseMode, Format
+            s.print()
+            fsink_config = {
+              'format': Format.txt.name,
+              'write_punctuations': True
+            }
+            script_dir = os.path.dirname(os.path.realpath(__file__))
+            debug_out_file = os.path.join(script_dir, 'flat_map.txt')
+            fsink = files.FileSink(file=debug_out_file, **fsink_config)
+            s.for_each(fsink)
+            self.test_config['topology.keepArtifacts'] = True
+
+        tester = Tester(topo)
+        tester.tuple_count(s, 10)
+        tester.test(self.test_ctxtype, self.test_config)
+
+    def test_object_in_object_out(self):
+        # Python object as input and output
+        # stream<blob __spl_po> GenerateNumbers = com.ibm.streamsx.topology.functional.python::Source  ( )
+        # stream<blob __spl_po> flatten_flat_map_obj = com.ibm.streamsx.topology.functional.python::FlatMap  ( GenerateNumbers)
+        topo = Topology("test_object_in_object_out")
+        s = topo.source(GenerateNumbers(2,5))
+        s = s.flat_map(flatten_flat_map_obj)
+        self._runTest(topo, s)
+
+    def test_object_in_namedtuple_out(self):
+        # Python object as input and named tuple as output
+        # stream<blob __spl_po> GenerateNumbers = com.ibm.streamsx.topology.functional.python::Source  ( )
+        # stream<int64 num> flatten_flat_map_obj_with_type_hints = com.ibm.streamsx.topology.functional.python::FlatMap  ( GenerateNumbers)
+        topo = Topology("test_object_in_namedtuple_out")
+        s = topo.source(GenerateNumbers(2,5))
+        s = s.flat_map(flatten_flat_map_obj_with_type_hints)
+        self._runTest(topo, s)
+
+    def test_dict_in_namedtuple_out(self):
+        # Python dict as input and named tuple as output
+        # stream<blob __spl_po> GenerateListOfDict = com.ibm.streamsx.topology.functional.python::Source  ( )
+        # stream<int64 num, rstring a, boolean b> flatten_flat_map_dict_with_type_hints = com.ibm.streamsx.topology.functional.python::FlatMap  ( GenerateListOfDict)
+        topo = Topology("test_dict_in_namedtuple_out")
+        s = topo.source(GenerateListOfDict(2,5))
+        s = s.flat_map(flatten_flat_map_dict_with_type_hints)
+        self._runTest(topo, s)
+
+    def test_namedtuple_in_namedtuple_out(self):
+        # named tuple as input and named tuple as output
+        # stream<list<int64> numlist> GenerateNumbersNamedTuple = com.ibm.streamsx.topology.functional.python::Source  ( )
+        # stream<int64 num> flatten_flat_map_with_type_hints = com.ibm.streamsx.topology.functional.python::FlatMap  ( GenerateNumbersNamedTuple)
+        topo = Topology("test_namedtuple_in_namedtuple_out")
+        s = topo.source(GenerateNumbersNamedTuple(2,5))
+        s = s.flat_map(flatten_flat_map_with_type_hints)
+        self._runTest(topo, s)
+
+    def test_namedtuple_in_object_out(self):
+        # named tuple as input and Python object as output
+        # stream<list<int64> numlist> GenerateNumbersNamedTuple = com.ibm.streamsx.topology.functional.python::Source  ( )
+        # stream<blob __spl_po> flatten_flat_map = com.ibm.streamsx.topology.functional.python::FlatMap  ( GenerateNumbersNamedTuple)
+        topo = Topology("test_namedtuple_in_object_out")
+        s = topo.source(GenerateNumbersNamedTuple(2,5))
+        s = s.flat_map(flatten_flat_map)
+        self._runTest(topo, s)
+
+
+class TestDistributedFlatMap(TestFlatMap):
+    def setUp(self):
+        Tester.setup_distributed(self)
+        self.test_config[ConfigParams.SSL_VERIFY] = False
+
+class TestSasFlatMap(TestFlatMap):
+    def setUp(self):
+        Tester.setup_streaming_analytics(self, force_remote_build=True)
+
+

